File: constraint-substrate/c/src/funnel.c
#include "constraint_substrate.h"

CsFunnelResult cs_funnel_step(double current, double target, double epsilon, double decay) {
    CsFunnelResult result;
    double diff = target - current;
    double exp_decay = exp(-decay);
    double new_eps = epsilon * exp_decay;

    if (fabs(diff) < epsilon) {
        /* Within deadband — snap toward target */
        double correction = diff * (1.0 - exp_decay);
        result.value = current + correction;
    } else {
        /* Outside deadband — pull toward target */
        double step_size = (diff > 0 ? 1.0 : -1.0) * epsilon;
        result.value = current + step_size;
    }
    result.epsilon = new_eps;
    return result;
}

File: constraint-substrate/python/constraint_substrate/funnel.py
"""Deadband convergence funnel."""

import math


def step(
    current: float, target: float, epsilon: float, decay_rate: float
) -> tuple[float, float]:
    """One step of deadband convergence with exponential decay.

    The epsilon (deadband width) decays exponentially:
        new_eps = epsilon * exp(-decay_rate)
    NOT linearly as epsilon * (1 - decay_rate).

    Exponential decay ensures:
    - Strictly positive epsilon (never reaches zero)
    - Smooth, natural convergence characteristic
    - Consistent halving behavior independent of current epsilon

    Returns (new_value, new_epsilon).
    """
    diff = target - current
    new_eps = epsilon * math.exp(-decay_rate)

    if abs(diff) < epsilon:
        # Within deadband — snap toward target
        correction = diff * (1.0 - math.exp(-decay_rate))
        return (current + correction, new_eps)
    else:
        # Outside deadband — pull toward target
        step_size = math.copysign(epsilon, diff)
        return (current + step_size, new_eps)


def step_batch(
    currents: list[float],
    targets: list[float],
    epsilons: list[float],
    decay: float,
) -> tuple[list[float], list[float]]:
    """Batch funnel step for N agents."""
    results = [step(c, t, e, decay) for c, t, e in zip(currents, targets, epsilons)]
    new_values = [r[0] for r in results]
    new_eps = [r[1] for r in results]
    return (new_values, new_eps)
